dartsim/src/EntityManagementFeatures.cc

```diff
@@ -272,22 +272,25 @@ std::size_t EntityManagementFeatures::GetLinkCount(
     const Identity &_modelID) const
 {
   return this->ReferenceInterface<ModelInfo>(_modelID)
-      ->model->getNumBodyNodes();
+      ->links.size();
 }
 
 /////////////////////////////////////////////////
 Identity EntityManagementFeatures::GetLink(
     const Identity &_modelID, const std::size_t _linkIndex) const
 {
-  DartBodyNode *const bn =
-      this->ReferenceInterface<ModelInfo>(_modelID)->model->getBodyNode(
-          _linkIndex);
+  auto modelInfo = this->ReferenceInterface<ModelInfo>(_modelID);
+
+  if (_linkIndex >= modelInfo->links.size())
+    return this->GenerateInvalidId();
+
+  auto linkInfo = modelInfo->links[_linkIndex];
 
   // If the link doesn't exist in "links", it means the containing entity has
   // been removed.
-  if (this->links.HasEntity(bn))
+  if (this->links.HasEntity(linkInfo->link))
   {
-    const std::size_t linkID = this->links.IdentityOf(bn);
+    const std::size_t linkID = this->links.IdentityOf(linkInfo->link);
     return this->GenerateIdentity(linkID, this->links.at(linkID));
   }
   else
```

Comment:
If the link has been removed, should we remove it from `this->links` before returning `this->GenerateInvalidId()`?

Reply:
Similar to my previous [comment](https://github.com/ignitionrobotics/ign-physics/pull/227#discussion_r598841185), I don't think we should have side effects in this function.